src/clone.rs

```diff
@@ -0,0 +1,99 @@
+use std::rc::{self, Rc};
+use std::sync::{self, Arc};
+
+pub trait Downgrade {
+    type Target;
+
+    fn downgrade(&self) -> Self::Target;
+}
+
+impl<T> Downgrade for Arc<T> {
+    type Target = sync::Weak<T>;
+
+    fn downgrade(&self) -> Self::Target {
+        Arc::downgrade(self)
+    }
+}
+
+impl<T> Downgrade for Rc<T> {
+    type Target = rc::Weak<T>;
+
+    fn downgrade(&self) -> Self::Target {
+        Rc::downgrade(self)
+    }
+}
+
+pub trait Upgrade {
+    type Target;
+
+    fn upgrade(&self) -> Option<Self::Target>;
+}
+
+impl<T> Upgrade for sync::Weak<T> {
+    type Target = Arc<T>;
+
+    fn upgrade(&self) -> Option<Self::Target> {
+        self.upgrade()
+    }
+}
+
+impl<T> Upgrade for rc::Weak<T> {
```

Comment:
These also need to be implemented in the `glib_wrapper` macro still

Reply:
This is done. Please do another round of review :) .